File: ROID-Forum-Web-Client/Convert Client Code/GMGWebsite/web/Website/Sections/GraphicsSection/ThreadViews/CommentView.dart
library GraphicsCommentView;
import '../ThreadInfo.dart';
import '../ThreadController.dart';
import '../CommentInfo.dart';
import 'dart:html';
import '../../../Tools/DescriptionParser.dart';

class CommentView {
  ThreadInfo thread;
  CommentInfo comment;
  ThreadController threadController;
  DivElement main, profileSection, descriptionSection, description;
  ImageElement image;
  TextAreaElement editDescription;
  ButtonElement editComment, saveEdit, deleteComment;
  SpanElement owner;
  CommentView(CommentInfo ci, ThreadInfo t, ThreadController tc) {
    threadController = tc;
    thread = t;
    comment = ci;
    makeMainComponent();
    makeImageComponent();
    makeOwnerComponent();
    makeDescriptionComponent();
    makeEditDescriptionComponent();
    makeEditCommentComponent();
    makeSaveEditComponent();
    makeDeleteCommentComponent();
  }
  void editButtonClick() {
    editComment.remove();
    descriptionSection.append(saveEdit);
    descriptionSection.append(deleteComment);
    descriptionSection.insertBefore(editDescription, description);
    description.remove();
    editDescription.value = comment.getComment();
  }
  void saveEditButtonClick() {
    descriptionSection.append(editComment);
    saveEdit.remove();
    deleteComment.remove();
    descriptionSection.insertBefore(description, editDescription);
    editDescription.remove();
    Map message = new Map();
    message['Controller'] = threadController.sectionController.name;
    message['Title'] = 'Edit Comment';
    message['Thread ID'] = thread.getID();
    message['Comment ID'] = comment.getCommentID();
    message['Text'] = editDescription.value;
    threadController.sectionController.website.networkingController.Send(message);
  }
  void deleteCommentButtonClick() {
    Map message = new Map();
    message['Controller'] = threadController.sectionController.name;
    message['Title'] = 'Delete Comment';
    message['Thread ID'] = thread.getID();
    message['Comment ID'] = comment.getCommentID();
    threadController.sectionController.website.networkingController.Send(message);
  }
  DivElement getDiv() {
    return main;
  }
  void updateOwner() {
    owner.text = comment.getPosterName();
    //If owner is self add edit button
    if (comment.getPosterName() == threadController.sectionController.website.database.name) {
      descriptionSection.append(editComment);
    }
  }
  void updateAvatarURL() {
      image.src = comment.getAvatarURL();
    }
  void updateDescription() {
    description.children.clear();
    description.append(DescriptionParser.parseDescription(comment.getComment()));
  }
  void makeMainComponent() {
    main = new DivElement();
    main.style..display = "inline-block"..marginLeft = "75px"..position="relative";
    main.style
        ..width = "calc(100% - 75px)"
        ..opacity = "1.0";

    profileSection = new DivElement();
    profileSection.style
        ..display = "inline-block"
        ..width = "60px"
        ..height = "80px"
        ..position = "absolute"
                ..top = "5px"
        ..textAlign = "center";
    main.append(profileSection);
    descriptionSection = new DivElement();
    descriptionSection.style
        ..display = "inline-block"
        ..width = "calc(100% - 95px)"
        ..marginLeft = "70px"
        ..marginTop = "5px"
        ..minHeight = "90px"
        ..textAlign = "left";
    main.append(descriptionSection);
    main.append(new HRElement());
  }
  void makeOwnerComponent() {
    owner = new SpanElement();
    owner.style
        ..fontSize = "15px"
        ..color = "black";
    profileSection.append(owner);
  }
  void makeImageComponent() {
    image = new ImageElement();
    image.style
        ..height = "60px"
        ..width = "60px"
        ..backgroundColor = "rgb(200,200,200)"
        ..borderStyle = "solid"
        ..borderWidth = "1px";
    profileSection.append(image);
  }
  void makeDescriptionComponent() {
    description = new DivElement();
    description.style
        ..width = "90%"
        ..wordWrap = "normal"
        ..color = "black";
    descriptionSection.append(description);
  }
  void makeEditDescriptionComponent() {
      editDescription = new TextAreaElement();
      editDescription.style
          ..width = "90%"
          ..wordWrap = "normal"
          ..color = "black";
    }
  void makeEditCommentComponent() {
    editComment = new ButtonElement();
    editComment.style.float = "right";
    editComment.text = "Edit";
    editComment.onClick.listen((_) {
      editButtonClick();
    });
  }
  void makeSaveEditComponent() {
    saveEdit = new ButtonElement();
    saveEdit.style.float = "right";
    saveEdit.text = "Save";
    saveEdit.onClick.listen((_) {
      saveEditButtonClick();
    });
  }
  void makeDeleteCommentComponent() {
    deleteComment = new ButtonElement();
    deleteComment.style.float = "left";
    deleteComment.text = "Delete";
    deleteComment.onClick.listen((_) {
      deleteCommentButtonClick();
    });
  }
}

File: ROID-Forum-Web-Client/src/Interface/CodingSection/ThreadViews/CommentView.ts
import {ThreadInfo} from "../ThreadInfo";
import {CommentInfo} from "../CommentInfo";
import {ThreadController} from "../ThreadController";
import {DescriptionParser} from "../../../Utility/DescriptionParser";
import {Utility} from "../../../Utility/Utility";
import {Interface} from "../../../Utility/Interface";

export class CommentView {
    thread: ThreadInfo;
    comment: CommentInfo;
    threadController: ThreadController;
    main: HTMLDivElement;
    profileSection: HTMLDivElement;
    descriptionSection: HTMLDivElement;
    description: HTMLDivElement;
    image: HTMLImageElement;
    editDescription: HTMLTextAreaElement;
    editComment: HTMLButtonElement;
    saveEdit: HTMLButtonElement;
    deleteComment: HTMLButtonElement;
    owner: HTMLSpanElement;

    constructor(ci: CommentInfo, t: ThreadInfo, tc: ThreadController) {
        this.threadController = tc;
        this.thread = t;
        this.comment = ci;
        this.makeMainComponent();
        this.makeImageComponent();
        this.makeOwnerComponent();
        this.makeDescriptionComponent();
        this.makeEditDescriptionComponent();
        this.makeEditCommentComponent();
        this.makeSaveEditComponent();
        this.makeDeleteCommentComponent();
    }

    editButtonClick() {
        this.editComment.remove();
        this.descriptionSection.appendChild(this.saveEdit);
        this.descriptionSection.appendChild(this.deleteComment);
        this.descriptionSection.insertBefore(this.editDescription, this.description);
        this.description.remove();
        this.editDescription.value = this.comment.getComment();
    }

    saveEditButtonClick() {
        this.descriptionSection.appendChild(this.editComment);
        this.saveEdit.remove();
        this.deleteComment.remove();
        this.descriptionSection.insertBefore(this.description, this.editDescription);
        this.editDescription.remove();
        let message: any = {};
        message['Controller'] = this.threadController.sectionController.name;
        message['Title'] = 'Edit Comment';
        message['Thread ID'] = this.thread.getID();
        message['Comment ID'] = this.comment.getCommentID();
        message['Text'] = this.editDescription.value;
        this.threadController.sectionController.website.networkingController.Send(message);
        console.log("Pressed Edit Comment on thread ${this.thread.getID()} and on comment ${this.comment.getCommentID()}");
    }

    deleteCommentButtonClick() {
        let message: any = {};
        message['Controller'] = this.threadController.sectionController.name;
        message['Title'] = 'Delete Comment';
        message['Thread ID'] = this.thread.getID();
        message['Comment ID'] = this.comment.getCommentID();
        this.threadController.sectionController.website.networkController.send(message);
    }

    getDiv(): HTMLDivElement {
        return this.main;
    }

    updateOwner() {
        this.owner.innerText = this.comment.getPosterName();
        //If owner is self add edit button
        if (this.comment.getPosterName() == this.threadController.sectionController.website.database.name) {
            this.descriptionSection.appendChild(this.editComment);
        }
    }

    updateAvatarURL() {
        this.image.src = this.comment.getAvatarURL();
    }

    updateDescription() {
        Utility.ClearElements(this.description);
        this.description.appendChild(DescriptionParser.parseDescription(this.comment.getComment()));
    }

    makeMainComponent() {
        this.main = Interface.Create({type: 'div'});
        this.main.style.display = "inline-block";
        this.main.style.marginLeft = "75px";
        this.main.style.position = "relative";
        this.main.style.width = "calc(100% - 75px)";
        this.main.style.opacity = "1.0";

        this.profileSection = Interface.Create({type: 'div'});
        this.profileSection.style.display = "inline-block";
        this.profileSection.style.width = "60px";
        this.profileSection.style.height = "80px";
        this.profileSection.style.position = "absolute";
        this.profileSection.style.top = "5px";
        this.profileSection.style.textAlign = "center";
        this.profileSection.style.color = "white";
        this.main.appendChild(this.profileSection);
        this.descriptionSection = Interface.Create({type: 'div'});
        this.descriptionSection.style.display = "inline-block";
        this.descriptionSection.style.width = "calc(100% - 95px)";
        this.descriptionSection.style.marginLeft = "70px";
        this.descriptionSection.style.marginTop = "5px";
        this.descriptionSection.style.minHeight = "90px";
        this.descriptionSection.style.textAlign = "left";
        this.descriptionSection.style.color = "white";
        this.main.appendChild(this.descriptionSection);
        this.main.appendChild(Interface.Create({type: 'hr'}));
    }

    makeOwnerComponent() {
        this.owner = Interface.Create({type: 'span'});
        this.owner.style.fontSize = "15px";
        this.owner.style.color = "white";
        this.profileSection.appendChild(this.owner);
    }

    makeImageComponent() {
        this.image = Interface.Create({type: 'img'});
        this.image.style.height = "60px";
        this.image.style.width = "60px";
        this.image.style.backgroundColor = "rgb(200,200,200)";
        this.image.style.borderStyle = "solid";
        this.image.style.borderWidth = "1px";
        this.profileSection.appendChild(this.image);
    }

    makeDescriptionComponent() {
        this.description = Interface.Create({type: 'div'});
        this.description.style.width = "90%";
        this.description.style.wordWrap = "normal";
        this.description.style.color = "white";
        this.descriptionSection.appendChild(this.description);
    }

    makeEditDescriptionComponent() {
        this.editDescription = Interface.Create({type: 'textarea'});
        this.editDescription.style.width = "90%";
        this.editDescription.style.wordWrap = "normal";
        this.editDescription.style.color = "black";
    }

    makeEditCommentComponent() {
        this.editComment = Interface.Create({type: 'button'});
        this.editComment.style.cssFloat = "right";
        this.editComment.innerText = "Edit";
        this.editComment.onclick = () => {
            this.editButtonClick();
        };
    }

    makeSaveEditComponent() {
        this.saveEdit = Interface.Create({type: 'button'});
        this.saveEdit.style.cssFloat = "right";
        this.saveEdit.innerText = "Save";
        this.saveEdit.onclick = () => {
            this.saveEditButtonClick();
        };
    }

    makeDeleteCommentComponent() {
        this.deleteComment = Interface.Create({type: 'button'});
        this.deleteComment.style.cssFloat = "left";
        this.deleteComment.innerText = "Delete";
        this.deleteComment.onclick = () => {
            this.deleteCommentButtonClick();
        };
    }
}